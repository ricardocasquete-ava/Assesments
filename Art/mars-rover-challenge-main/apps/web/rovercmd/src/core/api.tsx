import { baseUrl } from './apiConfig';

async function post<T> (url: string, data: Object): Promise<T> {
    const rawResponse = await fetch(`${baseUrl}${url}`, {
        method: 'POST',
        mode: 'cors',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
    });

    return rawResponse.json() as Promise<T>;
}

export { post };