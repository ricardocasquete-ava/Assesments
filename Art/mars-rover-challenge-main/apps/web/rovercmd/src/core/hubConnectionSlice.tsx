import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from './store';

interface HubConnectionState {
    connectionId: string | null
};

const initialState: HubConnectionState = {
    connectionId: null
};

export const hubConnectionSlice = createSlice({
  name: 'hubConnection',
  initialState,
  reducers: {
    setConnectionId: (state, action: PayloadAction<string | null>) => {
      state.connectionId = action.payload
    }
  },
});

export const { setConnectionId } = hubConnectionSlice.actions;

export const getConnectionId = (state: RootState) => state.hubConnection.connectionId;

export default hubConnectionSlice.reducer;