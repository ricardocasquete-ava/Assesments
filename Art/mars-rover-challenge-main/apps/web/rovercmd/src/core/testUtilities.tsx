import { render as reactRender, RenderResult } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import roverReducer from './roverSlice';
import hubConnectionReducer from './hubConnectionSlice';

function render(
  ui: React.ReactElement<any>,
  {
    preloadedState,
    store = configureStore({
      reducer: {
        rover: roverReducer,
        hubConnection: hubConnectionReducer
      },
      preloadedState
    }),
    ...renderOptions
  }: any = {}
): RenderResult {
  function Wrapper({ children }: any) {
    return <Provider store={store}>{children}</Provider>
  }
  return reactRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }