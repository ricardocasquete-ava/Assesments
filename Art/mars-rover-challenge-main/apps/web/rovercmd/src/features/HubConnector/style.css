.HubConnector {
  font-size: calc(10px + 0.5vmin);
  color: #a9a9a9;
}

.HubConnector a {
  color: #1e90ff;
  text-decoration: none;
  padding: 0px 5px;
}

.HubConnector a:hover {
  text-decoration: underline;
}