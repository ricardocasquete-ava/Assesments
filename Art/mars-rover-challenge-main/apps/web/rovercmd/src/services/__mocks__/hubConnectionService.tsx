export const getHubConnection = jest.fn();