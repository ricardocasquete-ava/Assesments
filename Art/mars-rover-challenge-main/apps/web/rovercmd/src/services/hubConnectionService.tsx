import { HubConnectionBuilder, HubConnectionState } from '@microsoft/signalr';
import { baseUrl } from 'core/apiConfig';

export interface RoverHubConnection {
    onclose: (callback: ((error?: Error | undefined) => void)) => void,
    onreconnecting: (callback: ((error?: Error | undefined) => void)) => void,
    onreconnected: (callback: ((connectionId: String | undefined) => void)) => void,
    on: (methodName: string, newMethod: (...args: any[]) => void) => void,

    start: () => Promise<void>,

    state: HubConnectionState,
    connectionId: string | null
}

export const getHubConnection = (): RoverHubConnection => {
    const newConnection = new HubConnectionBuilder()
        .withUrl(`${baseUrl}/hub`)
        .withAutomaticReconnect()
        .build();

    return newConnection;
}