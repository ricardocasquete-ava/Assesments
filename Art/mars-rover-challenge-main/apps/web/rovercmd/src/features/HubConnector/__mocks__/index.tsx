import 'features/HubConnector/style.css';

function HubConnector() {
  return (
    <div className="HubConnector">
      Mock Hub Connector
    </div>
  );
}

export default HubConnector;