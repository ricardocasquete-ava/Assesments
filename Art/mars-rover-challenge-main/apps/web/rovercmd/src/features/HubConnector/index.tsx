import { useState, useEffect } from 'react'
import { useAppSelector, useAppDispatch } from 'core/hooks';
import { setConnectionId, getConnectionId } from 'core/hubConnectionSlice';
import { HubConnectionState } from '@microsoft/signalr';
import { RoverHubConnection, getHubConnection } from 'services/hubConnectionService';
import './style.css';

function HubConnector() {
  const [ connection, setConnection ] = useState(null as unknown as RoverHubConnection);
  const [ status, setStatus ] = useState(HubConnectionState.Disconnected);
  const connectionId = useAppSelector(getConnectionId);
  const dispatch = useAppDispatch();

  useEffect(() => {
    initialiseHub();
  }, []);

  useEffect(() => {
    configureHub();
    connectToHub();
  }, [connection]);

  const initialiseHub = () => {
    const hubConnection = getHubConnection();
    setConnection(hubConnection);
  };

  const configureHub = () => {
    if (!connection) { return; }

    connection.onclose(updateStatus);
    connection.onreconnecting(updateStatus);
    connection.onreconnected(updateStatus);

    connection.on('ReceiveUpdate', (update: String) => {
      console?.log(`Received update! ${update}`);
    });
  };

  const connectToHub = () => {
    if (connection) {
      try {
        connection.start().finally(updateStatus);
      }
      catch {
        updateStatus();
      }
    }
  };

  const updateStatus = () => {
    if (connection) {
      setStatus(connection.state);
    }

    dispatch(setConnectionId(connection?.connectionId));
  };

  return (
    <div className="HubConnector" title={connectionId ? `id: ${connectionId}` : undefined}>
      {`${status}`}
      <a href='/' target='_self'
        hidden={status !== HubConnectionState.Disconnected}
        onClick={(e) => {
          e.preventDefault();
          connectToHub();
        }}>Reconnect?</a>
    </div>
  );
}

export default HubConnector;