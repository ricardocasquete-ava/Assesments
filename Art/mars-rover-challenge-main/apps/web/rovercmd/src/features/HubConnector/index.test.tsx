import { render } from 'core/testUtilities';
import { getHubConnection } from 'services/hubConnectionService';
import { HubConnectionState } from '@microsoft/signalr';
import { RoverHubConnection } from 'services/hubConnectionService';

jest.mock('services/hubConnectionService');

import HubConnector from './index';

let mockGetHubConnection: jest.Mock;
let mockHubConnection: RoverHubConnection;

beforeEach(() => {
  mockHubConnection = {
    onclose: jest.fn(),
    onreconnecting: jest.fn(),
    onreconnected: jest.fn(),
    on: jest.fn(),

    start: jest.fn(),

    state: HubConnectionState.Disconnected,
    connectionId: null
  }

  mockGetHubConnection = getHubConnection as jest.Mock;
  mockGetHubConnection.mockReturnValue(mockHubConnection);
});

test('renders initial HubConnector correctly', () => {
  const { container } = render(<HubConnector />);

  expect(container.firstChild).toMatchSnapshot();
});

test('renders HubConnector with active connection correctly', () => {
  mockHubConnection.state = HubConnectionState.Connected;
  mockHubConnection.connectionId = 'connectionId';
  const mockStart = mockHubConnection.start as jest.Mock;

  const { container } = render(<HubConnector />);

  mockStart.mockResolvedValue({});

  expect(container.firstChild).toMatchSnapshot();
});

test('renders HubConnector with reconnecting state', () => {
  mockHubConnection.state = HubConnectionState.Reconnecting;
  const mockStart = mockHubConnection.start as jest.Mock;

  const { container } = render(<HubConnector />);

  mockStart.mockResolvedValue({});

  expect(container.firstChild).toMatchSnapshot();
});

test('renders HubConnector with connection error', () => {
  mockHubConnection.state = HubConnectionState.Disconnected;
  const mockStart = mockHubConnection.start as jest.Mock;

  const { container } = render(<HubConnector />);

  mockStart.mockImplementation(() => { throw new Error("Error"); });

  expect(container.firstChild).toMatchSnapshot();
});
