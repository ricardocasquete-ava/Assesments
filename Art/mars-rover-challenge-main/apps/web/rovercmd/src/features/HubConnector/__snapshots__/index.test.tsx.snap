// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders HubConnector with active connection correctly 1`] = `
<div
  class="HubConnector"
  title="id: connectionId"
>
  Connected
  <a
    hidden=""
    href="/"
    target="_self"
  >
    Reconnect?
  </a>
</div>
`;

exports[`renders HubConnector with connection error 1`] = `
<div
  class="HubConnector"
>
  Disconnected
  <a
    href="/"
    target="_self"
  >
    Reconnect?
  </a>
</div>
`;

exports[`renders HubConnector with reconnecting state 1`] = `
<div
  class="HubConnector"
>
  Reconnecting
  <a
    hidden=""
    href="/"
    target="_self"
  >
    Reconnect?
  </a>
</div>
`;

exports[`renders initial HubConnector correctly 1`] = `
<div
  class="HubConnector"
>
  Disconnected
  <a
    href="/"
    target="_self"
  >
    Reconnect?
  </a>
</div>
`;
